ceph: Add support for volume replication controller sidecar in Rook (#7403)
rook/rook: design/ceph/ceph-csi-driver.md

#### Select Flex Provisioning or CSI based on CephCluster CRD

Currently the code requires changing numerous parameters to enable CSI. This
document aims to change that to a single parameter. In the future it may be
desirable to make this more of a "runtime" parameter that could be managed in
the cluster CRD.

### Support RBD Mirroring

> Start a Volume Replication sidecar container in the RBD Provisioner pod to start supporting RBD Mirroring between Ceph clusters.

Reviewer: more details need to be added here on how to do failover and failback for DR. refer https://github.com/kube-storage/volume-replication-operator#design

document aims to change that to a single parameter. In the future it may be
desirable to make this more of a "runtime" parameter that could be managed in
the cluster CRD.

### RBD Mirroring

To support RBD Mirroring, the [Volume Replication Operator](https://github.com/csi-addons/volume-replication-operator/blob/main/README.md) will be started in the RBD provisioner pod.
Volume Replication Operator is a kubernetes operator that provides common and reusable APIs for storage disaster recovery. It is based on [csi-addons/spec](https://github.com/csi-addons/spec) specification and can be used by any storage provider.
It follows controller pattern and provides extended APIs for storage disaster recovery. The extended APIs are provided via Custom Resource Definition (CRD).
To enable volume replication:

> - For Helm deployments see the [helm settings](https://github.com/rook/rook/blob/master/Documentation/helm-operator.md#configuration).

Reviewer 2: ```suggestion
- For Helm deployments see the [helm settings](helm-operator.md#configuration).
```

> - For non-Helm deployments set `CSI_ENABLE_VOLUME_REPLICATION: "true"` in the [operator.yaml](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml)

Reviewer 2: Actually let's remove this link to master
```suggestion
- For non-Helm deployments set `CSI_ENABLE_VOLUME_REPLICATION: "true"` in the operator.yaml
```

document aims to change that to a single parameter. In the future it may be
desirable to make this more of a "runtime" parameter that could be managed in
the cluster CRD.

> ### Support RBD Mirroring

Reviewer 2: ```suggestion
### RBD Mirroring
```

> Start [Volume Replication Operator](https://github.com/csi-addons/volume-replication-operator/blob/main/README.md) in the RBD provisioner pod.

Reviewer 2: ```suggestion
To support RBD Mirroring, the [Volume Replication Operator](https://github.com/csi-addons/volume-replication-operator/blob/main/README.md) will be started in the RBD provisioner pod.
```

> Volume Replication Operator is a kubernetes operator that provides common and reusable APIs for storage disaster recovery. It is based on[csi-addons/spec](https://github.com/csi-addons/spec) specification and can be used by any storage provider.

Reviewer: ```suggestion
Volume Replication Operator is a kubernetes operator that provides common and reusable APIs for storage disaster recovery. It is based on [csi-addons/spec](https://github.com/csi-addons/spec) specification and can be used by any storage provider.
```

> It follows controller pattern and provides extended APIs for storage disaster recovery. The extended APIs are provided via Custom Resource Definition (CRD).

Reviewer 2: Let's document how they would enable the replication controller, perhaps just to link to the helm doc and mention the env var in the operator.yaml if they aren't using helm
